File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void backtrack(string &digits, int index, unordered_map<char, string> &digitToLetters, string &current, vector<string> &result){
        if(index == digits.size()){
            result.push_back(current);
            return;
        }
        
        string letters = digitToLetters[digits[index]];
        for(int i=0; i<letters.size(); i++){
            current.push_back(letters[i]);
            backtrack(digits, index+1, digitToLetters, current, result);
            current.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        if(digits.size() == 0){
            return result;
        }
        
        unordered_map<char, string> digitToLetters = {{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
        
        string current;
        backtrack(digits, 0, digitToLetters, current, result);
        
        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void backtrack(String digits, int index, Map<Character, String> digitToLetters, StringBuilder current, List<String> result){
        if(index == digits.length()){
            result.add(current.toString());
            return;
        }
        
        String letters = digitToLetters.get(digits.charAt(index));
        for(char letter: letters.toCharArray()){
            current.append(letter);
            backtrack(digits, index+1, digitToLetters, current, result);
            current.deleteCharAt(current.length()-1);
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if(digits.length() == 0){
            return result;
        }
        
        Map<Character, String> digitToLetters = new HashMap<>();
        digitToLetters.put('2', "abc");
        digitToLetters.put('3', "def");
        digitToLetters.put('4', "ghi");
        digitToLetters.put('5', "jkl");
        digitToLetters.put('6', "mno");
        digitToLetters.put('7', "pqrs");
        digitToLetters.put('8', "tuv");
        digitToLetters.put('9', "wxyz");
        
        StringBuilder current = new StringBuilder();
        backtrack(digits, 0, digitToLetters, current, result);
        
        return result;
    }
}